[GTK] Make it possible to build with Skia instead of Cairo

carlosgcampos opened the pull request:
#### 75df56fbd143a4ac7df5e554b01c22030ea0ab22
<pre>
[GTK] Make it possible to build with Skia instead of Cairo
<a href="https://bugs.webkit.org/show_bug.cgi?id=270510">https://bugs.webkit.org/show_bug.cgi?id=270510</a>

Reviewed by Don Olmstead.

Add build option to enable Skia in the GTK port. This is just an initial
patch, there are many features broken when building with Skia like
printing, non-accelerated mode, favicons, snapshots, etc. We will fix
them all in follow up patches.

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/DragImage.h:
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
* Source/WebCore/platform/graphics/IntRect.h:
* Source/WebCore/platform/graphics/cairo/RefPtrCairo.cpp:
* Source/WebCore/platform/graphics/cairo/RefPtrCairo.h:
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.h:
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/gtk/CursorGtk.cpp:
(WebCore::createCustomCursor):
* Source/WebCore/platform/gtk/DragImageGtk.cpp:
(WebCore::dragImageSize):
(WebCore::scaleDragImage):
(WebCore::dissolveDragImageToFraction):
* Source/WebCore/platform/gtk/RenderThemeGadget.cpp:
* Source/WebCore/platform/gtk/RenderThemeGadget.h:
* Source/WebCore/platform/gtk/RenderThemeScrollbar.cpp:
* Source/WebCore/platform/gtk/RenderThemeScrollbar.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.cpp:
(API::ProcessPoolConfiguration::copy):
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseGetFaviconInternal):
(webkit_favicon_database_get_favicon_finish):
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(webkitWebContextConstructed):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::begin):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::setViewNeedsDisplay):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
* Source/WebKit/UIProcess/Automation/cairo/WebAutomationSessionCairo.cpp:
* Source/WebKit/UIProcess/BackingStore.h:
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.cpp:
(WebKit::DrawingAreaProxyCoordinatedGraphics::DrawingAreaProxyCoordinatedGraphics):
(WebKit::DrawingAreaProxyCoordinatedGraphics::setBackingStoreIsDiscardable):
(WebKit::DrawingAreaProxyCoordinatedGraphics::update):
(WebKit::DrawingAreaProxyCoordinatedGraphics::exitAcceleratedCompositingMode):
(WebKit::DrawingAreaProxyCoordinatedGraphics::enterAcceleratedCompositingMode):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.h:
* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
(WebKit::BackingStore::incorporateUpdate):
(WebKit::BackingStore::scroll):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::BufferSHM::didUpdateContents):
* Source/WebKit/UIProcess/gtk/ViewSnapshotStoreGtk3.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
(WebKit::WebPrintOperationGtk::startPrint):
(WebKit::WebPrintOperationGtk::rotatePageIfNeeded):
(WebKit::WebPrintOperationGtk::prepareContextToDraw):
(WebKit::WebPrintOperationGtk::renderPage):
(WebKit::WebPrintOperationGtk::printPagesDone):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp:
(WebKit::GtkSettingsManagerProxy::applyHintingSettings):
(WebKit::GtkSettingsManagerProxy::applyAntialiasSettings):
* Source/cmake/OptionsGTK.cmake:
* Tools/TestWebKitAPI/glib/WebKitGLib/TestMain.h:
* Tools/WebKitTestRunner/PlatformGTK.cmake:
* Tools/WebKitTestRunner/gtk/PlatformWebViewGtk.cpp:
(WTR::PlatformWebView::windowSnapshotImage):
* Tools/WebKitTestRunner/gtk/TestControllerGtk.cpp:
(WTR::TestController::takeViewPortSnapshot):

Canonical link: <a href="https://commits.webkit.org/275829@main">https://commits.webkit.org/275829@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f5ae0bed3a59d5322fd3c3a560daf521019ca51

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42991 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22017 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45393 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45615 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39116 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45297 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19440 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19049 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/37018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16641 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/38077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1047 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/36447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/38408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47145 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42618 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/17838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14692 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42325 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19443 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40977 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19622 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49628 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5819 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19073 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10029 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/WebCore/PlatformGTK.cmake:
```diff
@@ -44,6 +49,8 @@ list(APPEND WebCore_PRIVATE_FRAMEWORK_HEADERS
     accessibility/atspi/AccessibilityObjectAtspi.h
     accessibility/atspi/AccessibilityRootAtspi.h
 
+    platform/graphics/cairo/RefPtrCairo.h
```
This is already in `Cairo.cmake` so it should be in that `USE_SKIA` block below

donny-dont commented on Source/WebKit/UIProcess/BackingStore.h:
```diff
@@ -25,10 +25,11 @@
 
 #include <WebCore/IntSize.h>
 #include <WebCore/PlatformImage.h>
+#include <WebCore/RefPtrCairo.h>
```
This will break some of the porting I'm doing for other platform. Please move it to the `USE(CAIRO) || PLATFORM(GTK)` block.

donny-dont commented on Source/WebKit/UIProcess/BackingStore.h:
```diff
@@ -61,11 +63,11 @@ class BackingStore {
     float m_deviceScaleFactor { 1 };
-    WebCore::PlatformImagePtr m_surface;
-    WebCore::PlatformImagePtr m_scrollSurface;
+    RefPtr<cairo_surface_t> m_surface;
+    RefPtr<cairo_surface_t> m_scrollSurface;
```
Please do a `PLATFORM(GTK)` check here and don't remove the `WebCore::PlatformImagePtr`. This will break some of the porting I'm doing for other platforms to use Skia.

If possible just modify the `PlatformImagePtr` definition.

carlosgcampos replied:
I can't change `PlatformImagePtr` because it's `SkImage`, and that's what we want except here. GTK uses cairo for rendering non-accelerated contents.

carlosgcampos replied:
Much simpler!

donny-dont commented on Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
```diff
@@ -28,15 +28,18 @@
 #include "BackingStore.h"
 
-#if USE(CAIRO)
+#if USE(CAIRO) || PLATFORM(GTK)
```
I'm going to add a stub for a `BackingStoreSkia.cpp` just lay off this change at the moment.

carlosgcampos replied:
I still need this, because `USE(CAIRO)` could be false, but we still need to use this in the GTK port. We are not going to use `BackingStoreSkia.cpp`

donny-dont commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -86,6 +83,8 @@ else ()
     WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_RESOURCE_USAGE PRIVATE OFF)
 endif ()
 
+WEBKIT_OPTION_DEFINE(USE_SKIA "Whether to use Skia instead of Cairo." PRIVATE OFF)
```
Lets move this to a multi-port option and place it in `WebKitFeatures.cmake`. Remove the same define from `OptionsWPE.cmake`.

donny-dont commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -263,6 +262,14 @@ SET_AND_EXPOSE_TO_BUILD(USE_ATSPI 1)
+if (USE_SKIA)
+    SET_AND_EXPOSE_TO_BUILD(USE_CAIRO FALSE)
+else ()
+    find_package(Fontconfig 2.13.0 REQUIRED)
+    find_package(Freetype 2.9.0 REQUIRED)
```
I wouldn't move these here but keep them at the start. I'm 99% sure I'll be removing the `find_package` calls in Skia's CMake because those libraries aren't used on Windows.

carlosgcampos replied:
We could leave the find package calls there and call them depending on the target platform

donny-dont reviewed (changes requested):
Good start just some minor changes requested.

donny-dont commented on Source/WebKit/UIProcess/BackingStore.h:
```diff
@@ -25,21 +25,26 @@
 #pragma once
 
-#if USE(GRAPHICS_LAYER_WC) || (USE(CAIRO) && !PLATFORM(WPE))
+#if USE(GRAPHICS_LAYER_WC) || PLATFORM(GTK) || (USE(CAIRO) && !PLATFORM(WPE))
```
After thinking about this a bit more this could be simplified to just
```suggestion
#if !PLATFORM(WPE))
```

carlosgcampos committed:
[GTK] Make it possible to build with Skia instead of Cairo
https://bugs.webkit.org/show_bug.cgi?id=270510

Reviewed by Don Olmstead.

Add build option to enable Skia in the GTK port. This is just an initial
patch, there are many features broken when building with Skia like
printing, non-accelerated mode, favicons, snapshots, etc. We will fix
them all in follow up patches.

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/DragImage.h:
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
* Source/WebCore/platform/graphics/IntRect.h:
* Source/WebCore/platform/graphics/cairo/RefPtrCairo.cpp:
* Source/WebCore/platform/graphics/cairo/RefPtrCairo.h:
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.h:
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/gtk/CursorGtk.cpp:
(WebCore::createCustomCursor):
* Source/WebCore/platform/gtk/DragImageGtk.cpp:
(WebCore::dragImageSize):
(WebCore::scaleDragImage):
(WebCore::dissolveDragImageToFraction):
* Source/WebCore/platform/gtk/RenderThemeGadget.cpp:
* Source/WebCore/platform/gtk/RenderThemeGadget.h:
* Source/WebCore/platform/gtk/RenderThemeScrollbar.cpp:
* Source/WebCore/platform/gtk/RenderThemeScrollbar.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.cpp:
(API::ProcessPoolConfiguration::copy):
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseGetFaviconInternal):
(webkit_favicon_database_get_favicon_finish):
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(webkitWebContextConstructed):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::begin):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::setViewNeedsDisplay):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
* Source/WebKit/UIProcess/Automation/cairo/WebAutomationSessionCairo.cpp:
* Source/WebKit/UIProcess/BackingStore.h:
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.cpp:
(WebKit::DrawingAreaProxyCoordinatedGraphics::DrawingAreaProxyCoordinatedGraphics):
(WebKit::DrawingAreaProxyCoordinatedGraphics::setBackingStoreIsDiscardable):
(WebKit::DrawingAreaProxyCoordinatedGraphics::update):
(WebKit::DrawingAreaProxyCoordinatedGraphics::exitAcceleratedCompositingMode):
(WebKit::DrawingAreaProxyCoordinatedGraphics::enterAcceleratedCompositingMode):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.h:
* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
(WebKit::BackingStore::incorporateUpdate):
(WebKit::BackingStore::scroll):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::BufferSHM::didUpdateContents):
* Source/WebKit/UIProcess/gtk/ViewSnapshotStoreGtk3.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
(WebKit::WebPrintOperationGtk::startPrint):
(WebKit::WebPrintOperationGtk::rotatePageIfNeeded):
(WebKit::WebPrintOperationGtk::prepareContextToDraw):
(WebKit::WebPrintOperationGtk::renderPage):
(WebKit::WebPrintOperationGtk::printPagesDone):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp:
(WebKit::GtkSettingsManagerProxy::applyHintingSettings):
(WebKit::GtkSettingsManagerProxy::applyAntialiasSettings):
* Source/cmake/OptionsGTK.cmake:
* Tools/TestWebKitAPI/glib/WebKitGLib/TestMain.h:
* Tools/WebKitTestRunner/PlatformGTK.cmake:
* Tools/WebKitTestRunner/gtk/PlatformWebViewGtk.cpp:
(WTR::PlatformWebView::windowSnapshotImage):
* Tools/WebKitTestRunner/gtk/TestControllerGtk.cpp:
(WTR::TestController::takeViewPortSnapshot):

Canonical link: https://commits.webkit.org/275829@main
